### ISSUE TITLE
Atrous Convolutions for Graph Convs

### ISSUE THREAD

**Issue by rbharath** [deepchem/deepchem#616]
Team Paranoid Android at sv.ai/hackathon reports that atrous convolutions give large boosts to accuracy of graph convolution in DeepChem! I hope the members of this team see the message and contribute their implementions as a PR :-)

To other folks out there, check out atrous convolutions for your graph convs!

Possibly Relevant:
https://arxiv.org/abs/1606.00915

**Comment by lilleswing** [comment]
Is the idea to add filters from only n={2,3} distance from every atom?

**Comment by rbharath** [comment]
Yeah, I think that's the rough idea. Instead of having graph convolutional neighborhoods be just the direct bonded neighbors, you could make an atrous neighborhood of an atom which is all atoms two bonds out. This would enable the receptive fields of the molecules to communicate over much larger distances and could enable big performance improvements for large molecules. I'm pretty excited about the potential of atrous convolutions for biologic design (https://github.com/deepchem/deepbiologic)

**Comment by peastman** [comment]
I saw a talk recently on Message Passing Neural Networks (a generalization of graph convolutions) where they added an extra "pseudo-atom" to the molecule graph that was connected to every atom in the molecule.  They found that improved the results, since it created a shortcut for information to propagate quickly over long distances.  A somewhat similar idea?

**Comment by peastman** [comment]
They're sort of orthogonal ideas.  You can select which type of MPNN you want (graph convolutions are a special case of them).  And you can also decide what set of "extra connections" you want to include.  You can mix and match those however you want.

**Comment by rbharath** [comment]
We should build both and benchmark all the combinations :)

**Comment by lilleswing** [comment]
@peastman we over here had the same thoughts about the "fake" atom connected to everything.  Expect a PR coming up -- we wanted to combine it with the set2set but expect it in the next week.

**Comment by miaecle** [comment]
Yeah, I am pretty interested, will try to implement a prototype.

**Comment by lilleswing** [comment]
Excited this is coming back -- while thinking about it all we have to do is edit the graph topology for skip connections instead of one away connections in ConvMolFeaturizer.

**Comment by evanfeinberg** [comment]
In case it's useful, reiterating what I mentioned in a subgroup meeting
last year: one can take matrix powers of the NxN adjacency matrix "A" to do
atrous graph convolution. So for those graph convolutional models that use
the basic update (or some variation): h = W * A * x, one can instead just
do h = W * (A^p) * x, where x is the input feature matrix, p is an integer
representing just how atrous you want your graph convolution to be, and W
is a weight matrix. For example, a "1" at the (i, j) position of (A^2)
indicates that atom_i and atom_j are two bonds away from each other. This
simple formulation could lower the bar significantly for coding.

Anecdotally, I haven't seen significant improvements in performance with
atrous vs. normal graph conv, at least on Tox21 and PDBBind, though perhaps
with more hyperparam searching it's possible. Interesting to see what comes
out of this!

**Comment by rbharath** [comment]
Closing this old discussion. Possibly out of date for state of art on graph convs
